File: apps/manager-server/internal/service/codexinspection/xai_billing_zero.go
package codexinspection

import (
	"context"
	"encoding/base64"
	"encoding/binary"
	"strconv"
	"strings"
	"time"

	"github.com/seakee/cpa-manager-plus/apps/manager-server/internal/model"
	"github.com/seakee/cpa-manager-plus/apps/manager-server/internal/store"
)

const xaiWebBillingURL = "https://grok.com/grok_api_v2.GrokBuildBilling/GetGrokCreditsConfig"

// An omitted REST percentage alone remains unknown (#710). Only a complete,
// successful gRPC response for the same active period establishes implicit zero.
func isXAIValidatedZero(text, start, end string, now time.Time) bool {
	from, err := time.Parse(time.RFC3339Nano, start)
	if err != nil {
		return false
	}
	until, err := time.Parse(time.RFC3339Nano, end)
	if err != nil || now.Before(from) || !now.Before(until) || len(text) > 64*1024 {
		return false
	}
	encoded := strings.Join(strings.Fields(text), "")
	if len(encoded) == 0 || len(encoded)%4 != 0 {
		return false
	}
	// grpc-web-text can flush independently padded chunks, including trailers.
	var data []byte
	for len(encoded) > 0 {
		n := len(encoded)
		if i := strings.IndexByte(encoded, '='); i >= 0 {
			n = ((i / 4) + 1) * 4
		}
		chunk, err := base64.StdEncoding.Strict().DecodeString(encoded[:n])
		if err != nil {
			return false
		}
		data = append(data, chunk...)
		encoded = encoded[n:]
	}
	var payload []byte
	done := false
	for len(data) > 0 {
		if done || len(data) < 5 {
			return false
		}
		flag := data[0]
		size := uint64(binary.BigEndian.Uint32(data[1:5]))
		data = data[5:]
		if size > uint64(len(data)) {
			return false
		}
		frame := data[:int(size)]
		data = data[int(size):]
		if flag == 0 && payload == nil {
			payload = frame
		} else if flag == 128 && payload != nil {
			statuses := 0
			for _, line := range strings.Split(string(frame), "\r\n") {
				if strings.HasPrefix(strings.ToLower(line), "grpc-status:") {
					statuses++
					if strings.TrimSpace(line[len("grpc-status:"):]) != "0" {
						return false
					}
				}
			}
			if statuses != 1 {
				return false
			}
			done = true
		} else {
			return false
		}
	}
	if payload == nil || !done {
		return false
	}
	messages := map[string]bool{"1": true, "1.2": true, "1.3": true, "1.4": true, "1.5": true, "1.6": true, "1.7": true, "1.8": true, "1.12": true, "1.6.1": true, "1.6.2": true, "1.6.3": true, "1.8.2": true, "1.8.3": true, "1.6.3.2": true, "1.6.3.3": true}
	structural := map[string]bool{"1": true, "1.8": true, "1.8.1": true, "1.8.2": true, "1.8.3": true, "1.8.2.1": true, "1.8.3.1": true}
	seen := map[string]bool{}
	values := map[string]uint64{}
	var scan func([]byte, string) bool
	scan = func(b []byte, path string) bool {
		for len(b) > 0 {
			tag, n := binary.Uvarint(b)
			if n <= 0 {
				return false
			}
			b = b[n:]
			field, wire := tag>>3, tag&7
			if field == 0 || field > 536870911 {
				return false
			}
			key := strconv.FormatUint(field, 10)
			if path != "" {
				key = path + "." + key
			}
			if key == "1.1" || (messages[key] && wire != 2) {
				return false
			}
			if structural[key] {
				if seen[key] {
					return false
				}
				seen[key] = true
			}
			switch wire {
			case 0:
				value, n := binary.Uvarint(b)
				if n <= 0 {
					return false
				}
				b = b[n:]
				values[key] = value
			case 2:
				size, n := binary.Uvarint(b)
				if n <= 0 {
					return false
				}
				b = b[n:]
				if size > uint64(len(b)) {
					return false
				}
				if messages[key] && !scan(b[:int(size)], key) {
					return false
				}
				b = b[int(size):]
			case 1:
				if len(b) < 8 {
					return false
				}
				b = b[8:]
			default:
				// Any float could contradict implicit zero; never infer from it.
				return false
			}
		}
		return true
	}
	return scan(payload, "") && values["1.8.1"] == 2 && values["1.8.2.1"] == uint64(from.Unix()) && values["1.8.3.1"] == uint64(until.Unix())
}

func (s *Service) enrichXAIBillingZero(ctx context.Context, setup store.Setup, settings model.ManagerCodexInspectionConfig, item account, summary *xaiBillingSummary) {
	if summary == nil || summary.UsagePercent != nil || !summary.HasWeeklyData || len(summary.ProductUsage) > 0 {
		return
	}
	start, err := time.Parse(time.RFC3339Nano, summary.PeriodStart)
	if err != nil {
		return
	}
	end, err := time.Parse(time.RFC3339Nano, summary.PeriodEnd)
	if err != nil {
		return
	}
	now := time.Now()
	if now.Before(start) || !now.Before(end) {
		return
	}
	bounded, cancel := context.WithTimeout(ctx, 6*time.Second)
	defer cancel()
	settingsResponse, _, err := s.requestProviderBillingAt(bounded, setup, settings, item, xaiCLIChatProxyBaseURL+"/settings", map[string]string{"Authorization": "Bearer $TOKEN$", "x-xai-token-auth": "xai-grok-cli", "x-grok-client-version": xaiGrokVersion})
	if err != nil || settingsResponse.StatusCode != 200 {
		return
	}
	tier := readString(parseRecord(settingsResponse.Body), "subscription_tier_display")
	if tier != "SuperGrok" && tier != "SuperGrok Heavy" {
		return
	}
	response, _, err := s.requestProviderAPICallAt(bounded, setup, settings, item, "POST", xaiWebBillingURL, map[string]string{
		"Authorization": "Bearer $TOKEN$", "Content-Type": "application/grpc-web-text+proto", "Accept": "application/grpc-web-text+proto", "x-grpc-web": "1", "Origin": "https://grok.com",
	}, "AAAAAAIIAA==")
	if err == nil && response.StatusCode == 200 && isXAIValidatedZero(response.BodyText, summary.PeriodStart, summary.PeriodEnd, time.Now()) {
		zero := float64(0)
		summary.UsagePercent = &zero
		summary.UsagePercentSource = "grpc-implicit-zero"
	}
}

File: apps/web/src/utils/quota/xaiBillingZero.ts
/** A missing REST percentage is unknown. Only a complete, successful gRPC
 * response for the same active period can establish a proto3 implicit zero.
 * See CodexBar #3261/#3325 and CPA Manager Plus #710.
 */
export const XAI_WEB_BILLING_URL =
  'https://grok.com/grok_api_v2.GrokBuildBilling/GetGrokCreditsConfig';
export const XAI_WEB_BILLING_DATA = 'AAAAAAIIAA==';
export const XAI_WEB_BILLING_HEADERS = {
  Authorization: 'Bearer $TOKEN$',
  'Content-Type': 'application/grpc-web-text+proto',
  Accept: 'application/grpc-web-text+proto',
  'x-grpc-web': '1',
  Origin: 'https://grok.com',
};

// Only schema-declared messages are traversed; unknown bytes remain opaque.
const messages = new Set([
  '1',
  '1.2',
  '1.3',
  '1.4',
  '1.5',
  '1.6',
  '1.7',
  '1.8',
  '1.12',
  '1.6.1',
  '1.6.2',
  '1.6.3',
  '1.8.2',
  '1.8.3',
  '1.6.3.2',
  '1.6.3.3',
]);

export function isXaiValidatedZero(
  text: string,
  start: string | undefined,
  end: string | undefined,
  now = Date.now()
): boolean {
  try {
    const from = Date.parse(start ?? '');
    const until = Date.parse(end ?? '');
    if (!(from <= now && now < until) || text.length > 64 * 1024) return false;
    // grpc-web-text may flush multiple separately padded base64 chunks.
    const encoded = text.replace(/\s/g, '');
    if (!encoded || encoded.length % 4 || /[^A-Za-z0-9+/=]/.test(encoded)) return false;
    const chunks = encoded.match(/[A-Za-z0-9+/]+={0,2}/g) ?? [];
    if (
      chunks.join('') !== encoded ||
      chunks.some((chunk) => chunk.length % 4 || btoa(atob(chunk)) !== chunk)
    )
      return false;
    const bytes = Uint8Array.from(chunks.map((chunk) => atob(chunk)).join(''), (c) =>
      c.charCodeAt(0)
    );
    let offset = 0;
    let data: Uint8Array | undefined;
    let done = false;
    while (offset < bytes.length) {
      if (done || offset + 5 > bytes.length) return false;
      const flag = bytes[offset];
      const length = new DataView(bytes.buffer).getUint32(offset + 1);
      offset += 5;
      if (offset + length > bytes.length) return false;
      const payload = bytes.slice(offset, offset + length);
      offset += length;
      if (flag === 0 && !data) data = payload;
      else if (flag === 128 && data) {
        const trailer = new TextDecoder('utf-8', { fatal: true }).decode(payload);
        const statuses = trailer.split('\r\n').filter((line) => /^grpc-status:/i.test(line));
        if (statuses.length !== 1 || !/^grpc-status:\s*0\s*$/i.test(statuses[0])) return false;
        done = true;
      } else return false;
    }
    if (!data || !done) return false;
    const values = new Map<string, bigint>();
    const seen = new Set<string>();
    const scan = (buffer: Uint8Array, path: string) => {
      let i = 0;
      const varint = (): bigint => {
        let value = 0n;
        for (let shift = 0; shift < 70; shift += 7) {
          if (i >= buffer.length) throw Error('truncated');
          const b = buffer[i++];
          if (shift === 63 && b > 1) throw Error('overflow');
          value |= BigInt(b & 127) << BigInt(shift);
          if (b < 128) return value;
        }
        throw Error('overflow');
      };
      while (i < buffer.length) {
        const tag = varint();
        const field = tag >> 3n;
        const wire = Number(tag & 7n);
        if (field < 1n || field > 536870911n) throw Error('invalid tag');
        const key = path ? `${path}.${field}` : `${field}`;
        if (key === '1.1' || (messages.has(key) && wire !== 2))
          throw Error('unexpected field type');
        // Duplicate structural fields could hide a conflicting period/config.
        if (['1', '1.8', '1.8.1', '1.8.2', '1.8.3', '1.8.2.1', '1.8.3.1'].includes(key)) {
          if (seen.has(key)) throw Error('duplicate');
          seen.add(key);
        }
        if (wire === 0) values.set(key, varint());
        else if (wire === 2) {
          const length = varint();
          if (length > BigInt(buffer.length - i)) throw Error('truncated');
          const next = i + Number(length);
          if (messages.has(key)) scan(buffer.slice(i, next), key);
          i = next;
        } else if (wire === 1) i += 8;
        // Any float could contradict implicit zero. Never infer from it.
        else throw Error('not an implicit zero');
        if (i > buffer.length) throw Error('truncated');
      }
    };
    scan(data, '');
    return (
      values.get('1.8.1') === 2n &&
      values.get('1.8.2.1') === BigInt(Math.floor(from / 1000)) &&
      values.get('1.8.3.1') === BigInt(Math.floor(until / 1000))
    );
  } catch {
    return false;
  }
}
